### PR TITLE
New upgrade time QQ health check: add check_if_new_quorum_queue_replicas_have_finished_initial_sync

### PR DESCRIPTION
**Pull request by illotum**
## Proposed Changes

In our infrastructure the tool that decides which node to replace is abstracted and has no notion of what runs within an instance, nor can run any commands. It spawns a deploy worker and that is what will check_if_node_is_quorum_critical. This means the same worker also has to ensure all newly spawned replicas are synchronized before signaling process complete back. It is an unfortunate limitation, but for cases that require this level of isolation and security, a separate CLI tool is needed.

This PR adds internal API and a diagnostics/queues command to check if the node has no promotable QQ replicas.

Some notable observations:
- I took liberty to add amqqueue:to_printable. The code it runs is now used in two places, but will happily revert.
- The command is check_if_node_has_promotable_replicas. It's the clearest I could come up with, suggestions welcome.
- There a fair bit of new user-facing text in the new command, would appreciate feedback regarding its clarity, etc.
- Unlike check_if_node_is_quorum_critical, this test doesn't handle streams at all, hope it is acceptable and streams can be added later.
- Nor does it have an equivalent HTTP API. Don't want to blow up the diff, but can add it as well.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [ ] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [ ] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

### CODE REVIEW

**Comment by michaelklishin** [comment]
This command was introduced specifically to verify if a node is safe to shut down.

We either need a new command or perhaps make the current one take the log installation (syncing) status into account?

Most users should roughly understand what `rabbitmq-queues check_if_node_is_quorum_critical` means but what does

``` shell
rabbitmq-queues check_if_node_is_quorum_critical --global
```

means… so the version without the flag is "local"? What exactly does that mean?

**Comment by michaelklishin** [comment]
If the goal is to check if the target node may be hosting any QQ replicas that are currently not considered votes because of their ongoing log installation (initial log sync), I'd suggest
something like

``` shell
rabbitmq-queues check_if_node_has_non_voting_quorum_queue_replicas
```

I am struggling to come up with a better adjective than "non-voting" but "incomplete", "out of sync" are both even less specific and can even be misleading.

**Comment by illotum** [comment]
That's absolutely fair. Back to draft for now.

**Comment by michaelklishin** [comment]
@illotum thank you. Should we name this `promotable_to_voter` or something more specific like that? "Promotable" in mean different things to different people, even strictly in the context of Raft, e.g. "promotable to a leader" (any replica with an up-to-date log is).

Or something like `check_if_node_has_non_voter_replicas`?

**Comment by illotum** [comment]
non_voter unfortunately has other meaning within Ra -- a permanent non_voter, which obviously should not be a cause for alarm.
How about `check_if_new(_quorum)_replicas_are_progressing`, or `_in_sync`?

**Comment by michaelklishin** [comment]
@illotum if you ask me, any online replica is always progressing and can be out of sync for a fraction of a second.

`check_if_new_quorum_replicas_have_finished_initial_sync` is the best I could come up with. I don't mind long command names if they are more specific :P

**Comment by illotum** [comment]
I put "new" there to indicate it is for newly added replicas. But I'm good with any name truth be told, renamed it to `check_if_new_quorum_replicas_have_finished_initial_sync` and rebased.

**Comment by michaelklishin** [comment]
@illotum thank you. I'd use `check_if_new_quorum_queue_replicas_have_finished_initial_sync` because "quorum replica" is not really a term we use anywhere.

**Review by kjnilsson — CHANGES_REQUESTED** [review]
A few minor comments as well as general feel that we don't need 3 new public functions in rabbit_quorum_queue to achieve this.

File: deps/rabbit/src/rabbit_quorum_queue.erl [hunk 1]
```diff
@@ -404,46 +408,48 @@ all_replica_states() ->
...
 
--spec filter_quorum_critical([amqqueue:amqqueue()], #{node() => #{atom() => atom()}}, node()) ->
+-spec list_with_local_promotable() -> [amqqueue:amqqueue()].
+list_with_local_promotable() ->
+    Queues = rabbit_amqqueue:list_local_quorum_queues(),
+    #{node() := ReplicaStates} = get_replica_states([node()]),
+    filter_promotable(Queues, ReplicaStates).
+
+-spec list_with_local_promotable_for_cli() -> [#{binary() => any()}].
+list_with_local_promotable_for_cli() ->
```
[the remark this answers is not in the document]
Comment (michaelklishin) [reply deps/rabbit/src/rabbit_quorum_queue.erl:422]: This is an established pattern in CLI tools. The server does the formatting, CLI tools accept a map or a list of maps that is easy to format and that does not depend on the server version (or at least does not depend on it much).

[the remark this answers is not in the document]
Comment (michaelklishin) [reply deps/rabbit/src/rabbit_quorum_queue.erl:422]: More specifically, we do not want CLI tools getting a record. Not only records are fragile in Erlang, in Elixir you have to explicitly hook them up from a header. We do that in a few places already but with very stable records (like `listener`).

File: deps/rabbit/src/rabbit_quorum_queue.erl [hunk 2]
```diff
@@ -404,46 +408,48 @@ all_replica_states() ->
...
                     {Name, _Node} = amqqueue:get_pid(Q),
                     AllUp = lists:filter(fun (N) ->
-                                            case maps:get(N, ReplicaStates, undefined) of
+                                            case maps:get(N, AllReplicaStates, undefined) of
                                                 #{Name := State}
-                                                  when State =:= follower orelse
-                                                       State =:= leader orelse
-                                                       (State =:= promotable andalso N =:= Self) orelse
-                                                       (State =:= non_voter andalso N =:= Self) ->
+                                                  when State == follower;
```
[the remark this answers is not in the document]
Comment (SimonUnge) [reply deps/rabbit/src/rabbit_quorum_queue.erl:449]: Agree. Big fan of `andalso` `orelse` over `,` and `;`

File: deps/rabbit/src/rabbit_quorum_queue.erl [hunk 3]
```diff
@@ -55,7 +55,9 @@
          file_handle_other_reservation/0]).
 -export([file_handle_release_reservation/0]).
 -export([list_with_minimum_quorum/0,
-         filter_quorum_critical/1,
+         list_with_local_promotable/0,
+         list_with_local_promotable_for_cli/0,
+         filter_promotable/2,
```
[the remark this answers is not in the document]
Comment (illotum) [reply deps/rabbit/src/rabbit_quorum_queue.erl:60]: filter_quorum_critical is. But sure, I'll embed the test that uses it.

File: deps/rabbit/src/rabbit_quorum_queue.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (illotum) [reply deps/rabbit/src/rabbit_quorum_queue.erl:425]: I take it as "keep as-is" then? If I weren't following existing patterns I'd probably default to Karl's approach, but there was a pattern indeed.

[the remark this answers is not in the document]
Comment (michaelklishin) [reply deps/rabbit/src/rabbit_quorum_queue.erl:425]: Yes, let's keep this because dealing with records in CLI tools is not worth the gain

**Pull request merged**